lock-api/src/test/java/com/palantir/lock/client/RemoteTimeoutRetryerTest.java

```diff
@@ -0,0 +1,164 @@
+ * limitations under the License.
+ */
+
+package com.palantir.lock.client;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+
+import java.net.SocketTimeoutException;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.function.Function;
+
+import org.immutables.value.Value;
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+
+@SuppressWarnings("unchecked") // Mocks with generic types
+public class RemoteTimeoutRetryerTest {
+    private static final int DEFAULT_VALUE = -1;
+    private static final int QUERY_VALUE = 3;
+    private static final int FORTY_TWO = 42;
+    private static final Exception TIMEOUT_EXCEPTION = new RuntimeException(new SocketTimeoutException("timeout"));
+    private static final Exception RUNTIME_EXCEPTION = new RuntimeException();
+
+    private final Clock clock = mock(Clock.class);
+    private final RemoteTimeoutRetryer retryer = new RemoteTimeoutRetryer(clock);
+    private final Function<IntAndDuration, Integer> query = mock(Function.class);
+
+    @Test
+    public void returnsRequestImmediatelyIfSuccessful() {
+        setupDefaultClock();
+        when(query.apply(any())).thenReturn(FORTY_TWO);
+
+        int result = retryForTenMillisAcceptingOnlyFortyTwo();
+        assertThat(result).isEqualTo(FORTY_TWO);
+
+        verify(query).apply(any(IntAndDuration.class));
+        verifyNoMoreInteractions(query);
+    }
+
+    @Test
+    public void returnsDefaultIfAlwaysFailing() {
+        setupDefaultClock();
+        when(query.apply(any())).thenReturn(5);
+
+        int result = retryForTenMillisAcceptingOnlyFortyTwo();
+        assertThat(result).isEqualTo(DEFAULT_VALUE);
+
+        verify(query, times(3)).apply(any(IntAndDuration.class));
+        verifyNoMoreInteractions(query);
+    }
+
+    @Test
+    public void blockingDurationsDecreaseAsTimePasses() {
+        setupDefaultClock();
+        when(query.apply(any())).thenReturn(5);
+
+        retryForTenMillisAcceptingOnlyFortyTwo();
+
+        ArgumentCaptor<IntAndDuration> captor = ArgumentCaptor.forClass(IntAndDuration.class);
+        verify(query, times(3)).apply(captor.capture());
+        assertThat(captor.getAllValues()).containsExactly(
+                ImmutableIntAndDuration.of(QUERY_VALUE, Duration.ofMillis(10)),
+                ImmutableIntAndDuration.of(QUERY_VALUE, Duration.ofMillis(6)),
+                ImmutableIntAndDuration.of(QUERY_VALUE, Duration.ofMillis(2)));
+        verifyNoMoreInteractions(query);
+    }
+
+    @Test
+    public void recoversFromFailureIfWeBecomeSuccessful() {
+        setupDefaultClock();
```

Comment:
the tests here are weird because there is nothing tying together a request timing out and the actual time, one is handled by the `setupDefaultClock` and the other is just if the query returns a correct response, as such it makes this example a bti contrived and hard to follow.

If anything, I'd probably have a method which was like: `returnResponseAfterDuration(T response, Duration duration)`. That sets up the clock and also the query mock.

```java
private static BoolAndDuration successfulAfter(Duration duration) { ... }
private static BoolAndDuration failedAfter(Duration duration) { ... }
private static void setupStuff(BoolAndDuration... durations) {
  // create arrays of the flag and stick it into the query mock
  // create arrays of the duration and stick it into the clock mock
}

public void test() {
  setupStuff(
    failedAfter(Duration.ofMillis(4)), 
    successfullAfter(Duration.ofMillis(5));

  bool result = retryForTenMillisWaitingForSuccess();
  ...
}
```

I'd probably go further and get rid of ints and just do this purely on booleans.

Reply:
The BDD style thing could work. I think this is somewhat more complicated than what you've written above though: there are 4 cases, not 2 (retryable and non-retryable exceptions) so we can't just use `thenReturn`, there'll be some tricky conversions involved.